lib/rspec/core.rb

```diff
@@ -31,6 +31,7 @@
   option_parser
   configuration_options
   runner
+  profiler
```

Comment:
> What you're saying makes sense, but reveals that I totally misunderstood your request on the other PR to rename the thing form `ExceptionFormatter` to something else.  You're right that the things we register with the reporter are "listeners" (not necessarily output formatters) but unfortunately, our public API speaks only of formatters. The public API is `add_formatter`, not `add_listener`.  It might be worth changing that in RSpec 4 (since registering a listener is the generic way to be notified of events...and it's always bothered me that a user has to register a "formatter" for that...) to `add_listener`.  Although, that might not be worth the pain it could cause for users...who knows.

The reporter does have `add_listener`, `add_formatter` does something else, it sets up output streams, instantiates a formatter and then registers it as a listener on the reporter. It's possible already for extensions to utilise that hook without going through the process of registering a formatter, the reason we do that for formatters is they need a way to be distinguished from legacy formatters and it makes it easier to specify them from the command line etc.


> I've been thinking of the "formatters" directory as the place where all the listener class definitions are, but we just happen to keep calling them "formatters" since that's what they've been called for years.  It feels really odd to me to split where the listener classes are defined (since they feel natural to be grouped to me) but you're right that it's also odd to put a non-formatting listener in `formatters`.  I wish the directory was called "listeners", frankly :(.

I don't, it'd be ok with having two directories though.

> We've been treating the `profile_examples` option in the way I'm suggesting for all RSpec 3 release -- after all, the [ProfileFormatter is only registered if `profile_examples` is set when `setup_default` is executing](https://github.com/rspec/rspec-core/blob/5375dcdfdb5460a51d5003de76eb124997fa26ea/lib/rspec/core/formatters.rb#L124-L126).  If the `ProfileFormatter` is registered, it's essential that this new listener is also registered, right?  Otherwise the `ProfileFormatter` won't work properly.  On the flip side, if the `ProfileFormatter` isn't registered, then this new listener serves no purpose, and it's wasted computation to register it and have it listen to the events.

> So why not register them both together at the same time based on the value of the config option at that point?

This I could do.

> I don't understand what the two way communication issue you refer to is or why this would necessitate moving it back onto the config object (where it doesn't belong, IMO).

Theres no two way communication required as long as you're ok with:

 ```Ruby
config.profile_examples = true
config.profile_examples = false
```

Not actually removing the profiler or the profile formatter.

> One thing that might help is to consider having the reporter always initialize (and expose) a `profile_results` hash.  When instantiating the Profile listener we can then pass that to it so that it can populate it as it listens.  That way, the reporter always has a valid `profile_results` object that it can provide to the notification regardless of whether or not the profiler listener was actually used.  It's just that it'll be empty if the profile listener wasn't used, which makes sense, IMO.

Thats what NoProfiler was for, I didn't want to conflate the reporter with responsibility it didn't need.

> Not following you here again.  I thought we did away with `NoProfiler`?  Why does this necessitate bringing that back?

It's a null object version of the profiler... if you have a way to turn profiling off you need one or you have to add if statements or unrelated state to the reporter, both of which I'd rather not do.

> Sorry if it seems like I'm being argumentative here...I'm not intending to be.  I just care a lot about organizing RSpec's code in such a way that for optional features like the profiling, users only pay the runtime cost of that extra computation if they've opted-in to use the feature (kinda like how you care so much about stdlib usage!).  I think such an approach is essential to keeping RSpec fast even as we add and refine features.

I care too which is why it seems like we're being argumentative ;)

> Maybe what I'm asking for isn't easily possible for a reason I'm not yet seeing.  But I'd like to at least understand why if that's the case.